File: Haskell/DataStructures/Heap/MaxiphobicHeap.hs
--------------------------------------------------------------------------------
-- Student's name: Alejandro Garau Madrigal ------------------------------------
--------------------------------------------------------------------------------
-- Maxiphobic Heaps ------------------------------------------------------------
-- Data Structures. Grado en Informática. UMA. ---------------------------------
--------------------------------------------------------------------------------

module DataStructures.Heap.MaxiphobicHeap
  ( Heap
  , empty
  , isEmpty
  , minElem
  , delMin
  , insert
  , merge
  , mkHeap
  ) where

import Test.QuickCheck

data Heap a  = Empty | Node a Int (Heap a) (Heap a) deriving Show

-- number of elements in tree rooted at node
weight :: Heap a -> Int
weight Empty           = 0
weight (Node _ w _ _)  = w


singleton :: a -> Heap a
singleton x  = Node x 1 Empty Empty


empty :: Heap a
empty = Empty


isEmpty :: Heap a -> Bool
isEmpty Empty = True
isEmpty _     = False


insert :: (Ord a) => a -> Heap a -> Heap a
insert it h = merge (singleton it) h


minElem :: Heap a -> a
minElem Empty          = error "minElem on empty Heap"
minElem (Node x _ _ _) = x


delMin :: (Ord a) => Heap a -> Heap a
delMin Empty            = error "delMin on empty heap"
delMin (Node _ _ lh rh) = merge lh rh


merge :: (Ord a) => Heap a -> Heap a -> Heap a
merge Empty h'  = h'
merge h Empty   = h
merge h@(Node x w lh rh) h'@(Node x' w' lh' rh')
    | x <= x'   = check (Node x s h1 (merge h2 h3))
    | otherwise = check (Node x' s' h1' (merge h2' h3'))
        where
            (h1, h2, h3, s) = node lh rh h'
            (h1', h2', h3', s') = node lh' rh' h

-- Efficient O(n) bottom-up construction for heaps
mkHeap :: (Ord a) => [a] -> Heap a
mkHeap []  = empty
mkHeap xs  = mergeLoop (map singleton xs)
  where
    mergeLoop [h]  = h
    mergeLoop hs   = mergeLoop (mergePairs hs)

    mergePairs []         = []
    mergePairs [h]        = [h]
    mergePairs (h:h':hs)  = merge h h' : mergePairs hs

--------------------------------------------------------------------------------
-- Private definitions ---------------------------------------------------------
--------------------------------------------------------------------------------
node :: Heap a -> Heap a -> Heap a -> (Heap a, Heap a, Heap a, Int)
node lh rh h
    | wh >= wrh && wrh >= wlh = (h, rh, lh, wh+1)
    | wlh >= wrh && wrh >= wh = (lh, rh, h, wlh+1)
    | otherwise               = (rh, lh, h, wrh+1)
        where
            wlh = weight lh
            wrh = weight rh
            wh = weight h

check :: Heap a -> Heap a
check Empty = Empty
check (Node x w lh rh) = if wl < wr then (Node x w rh lh) else (Node x w lh rh)
    where
        wl = weight lh
        wr = weight rh

--------------------------------------------------------------------------------
-- Generating arbritray Heaps --------------------------------------------------
--------------------------------------------------------------------------------

instance (Ord a, Arbitrary a) => Arbitrary (Heap a) where
  arbitrary  = do
    xs <- arbitrary
    return (mkHeap xs)

a :: Heap Int
a = (Node 5 2) ((Node 6 0) Empty Empty) ((Node 7 0) Empty Empty)

b :: Heap Int
b = (Node 8 2) (Node 10 0 Empty Empty) (Node 15 0 Empty Empty)

File: java/MaxiphobicHeap/src/dataStructures/MaxiphobicHeap/MaxiphobicHeap.java
/**
 * @author Alejandro Garau Madrigal
 * 
 * Maxiphobic Heaps
 * Data Structures, Grado en Informática. UMA.
 *
 */
 
package dataStructures.MaxiphobicHeap;


/**
 * Heap implemented using maxiphobic heap-ordered binary trees.
 * @param <T> Type of elements in heap.
 */
public class MaxiphobicHeap<T extends Comparable<? super T>> implements	Heap<T> {

	protected static class Tree<E> {
		E elem;
		int weight;
		Tree<E> left;
		Tree<E> right;
	}

	private static int weight(Tree<?> heap) {
		return heap == null ? 0 : heap.weight;
	}

	private static <T extends Comparable<? super T>> Tree<T> merge(Tree<T> h1,	Tree<T> h2) {
        if (h1 == null)
            return h2;
        if (h2 == null)
            return h1;

        if(h1.elem.compareTo(h2.elem) > 0) {
            Tree<T> aux = h1;
            h1 = h2;
            h2 = aux;
        }
        /*Tree<T> ret = new Tree<>();
        Tupla<T> tup = node(h1.left, h1.right, h2);
        ret.elem = h1.elem;
        ret.left = tup.h;
        ret.right = merge(tup.lh, tup.rh);
        ret.weight = weight(h1) + weight(h2);
        return ret;*/
        Tree<T> tree = new Tree<>();
        tree.elem = h1.elem;
		// int max = Math.max(Math.max(weight(h1.left), weight(h1.right)), weight(h2));
		int wh1l = weight(h1.left);
		int wh1r = weight(h1.right);
		int wh2  = weight(h2);
        if(wh2 >= wh1l && wh2 >= wh1r){
            tree.weight = weight(h2) + 1;
            tree.left = h2;
            tree.right = merge(h1.left, h1.right);
        } else if (wh1l >= wh2 && wh1l >= wh1r){
            tree.weight = weight(h1.left) + 1;
            tree.left = h1.left;
            tree.right = merge(h1.right, h2);
        }
        else{
			tree.weight = weight(h1.right) + 1;
			tree.left = h1.right;
			tree.right = merge(h1.left, h2);
		}
		/*COMPROBAR QUE SEA ZURDO*/
		if(weight(tree.right) > weight(tree.left)){
            Tree<T> aux = tree.left;
            tree.left = tree.right;
            tree.right = aux;
        }

        return tree;
	}

    protected Tree<T> root;

	/**
	 * Constructor: creates an empty Maxiphobic Heap.
	 * <p>Time complexity: O(1)
	 */
	public MaxiphobicHeap() {
		this.root = null;
	}

	/** 
	 * <p>Time complexity: O(1)
	 */
	public boolean isEmpty() {
		return this.root == null;
	}

    /**
     * @return size
     * <p>Time complexity: O(1)
     */
	@Override
	public int size() {
		return isEmpty() ? 0 : root.weight;
	}

	/** 
	 * <p>Time complexity: O(1)
     * @throws EmptyHeapException if heap stores no element.
	 */
	public T minElem() {
		if (isEmpty())
		    throw new EmptyHeapException("minElem on empty Heap");
		return root.elem;
	}

	/** 
	 * <p>Time complexity: O(log n)
	 * @throws EmptyHeapException if heap stores no element.
	 */
	public void delMin() {
		if(isEmpty())
		    throw new EmptyHeapException("delMin on empty Heap.");
		root = merge(root.left, root.right);
	}

	/** 
	 * <p>Time complexity: O(log n)
	 */
	public void insert(T value) {
		Tree<T> item = new Tree<>();
		item.elem = value;
		item.weight = 1;
		item.right = item.left = null;
		root = merge(item, root);
	}

	private static String toStringRec(Tree<?> tree) {
		return tree == null ? "null" : "Node<" + toStringRec(tree.left) + ","
				+ tree.elem + "," + toStringRec(tree.right) + ">";
	}
	
	/** 
	 * Returns representation of heap as a String.
	 */
	public String toString() {
		String className = getClass().getSimpleName();

		return className+"("+toStringRec(this.root)+")";
	}	
	
}